prov/hook/profile/src/prof_report.c

```diff
@@ -0,0 +1,251 @@
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include <stdio.h>
+#include <string.h>
+
+#include "ofi_str.h"
+#include "hook_profile.h"
+
+#define PROF_STR_LEN	64
+#define PROF_HMEM_IFACE_MAX	  FI_HMEM_SYNAPSEAI+1
+
+#define PROF_OUTPUT_FORMAT    " \t%-22s%-20s%-12s%-12s%-12s%-12s\n"
+
+static const char *prof_api_name[] = {
+	PROFILE_APIS(OFI_STR)
+};
+
+static char prof_api_disp_str[prof_api_size][PROF_STR_LEN];
+static char prof_size_str[prof_api_size][PROF_STR_LEN];
+
+static bool prof_disp_name_avail = false;
+
+/* get suffix str,  from PROF_SIZE_A_B  to "A_B" */
+#define CASEENUMSUFFIX(SYM, prefix, len)  \
+    case SYM:  { char *symstr = #SYM;  \
+		         ofi_strncatf(buf, len, &symstr[prefix]); break; }
+
+/* convert "prof_<ops>"  to  "fi_<ops>" for log name*/
+static inline void
+set_api_name(char *buf, size_t size, const char *api_str) {
+	int len = strlen("prof_");
+	ofi_strncatf(buf, size, "fi_");
+	ofi_strncatf(buf, size, &api_str[len]);
+}
+
+/* convert "prof_cq_xxx_tx"  to cq_xxx (FI_SEND)
+ * convert "prof_cq_xxx_rx"  to cq_xxx (FI_RECV)
+ */
+static inline void
+set_cq_api_name(char *buf, size_t size, const char *api_str) {
+	int len = strlen(api_str);
+	int prefix = strlen("prof_");
+
+	ofi_strncatf(buf, size, &api_str[prefix]);
+	buf[strlen(buf)-3] = '\0';
+	if (!strncmp(&api_str[len-3], "_tx", 3))
+		ofi_strncatf(buf, size, " (FI_SEND)");
+	else
+		ofi_strncatf(buf, size, " (FI_RECV)");
+}
+
+// string name for enum prof_size_bucket
+static char *prof_size_bucket_tostr(char *buf, size_t buf_size,
+                                    enum prof_size_bucket bucket) {
+	int prefix_len = strlen("PROF_SIZE_");
+	buf[0] = '\0';
+	switch(bucket) {
+	CASEENUMSUFFIX(PROF_SIZE_0_64, prefix_len, buf_size);
+	CASEENUMSUFFIX(PROF_SIZE_64_512, prefix_len, buf_size);
+	CASEENUMSUFFIX(PROF_SIZE_512_1K, prefix_len, buf_size);
+	CASEENUMSUFFIX(PROF_SIZE_1K_4K, prefix_len, buf_size);
+	CASEENUMSUFFIX(PROF_SIZE_4K_64K, prefix_len, buf_size);
+	CASEENUMSUFFIX(PROF_SIZE_64K_256K, prefix_len, buf_size);
+	CASEENUMSUFFIX(PROF_SIZE_256K_1M, prefix_len, buf_size);
+	CASEENUMSUFFIX(PROF_SIZE_1M_4M, prefix_len, buf_size);
+	CASEENUMSUFFIX(PROF_SIZE_4M_Up, prefix_len, buf_size);
+	default:
+		break;
+    }
+    return buf;
+}
+
+
+static char *ratio_to_str(char *str, int len, uint64_t value, uint64_t total)
+{
+	snprintf(str, len, "%6.2f", 100.0 * value / total);
+	return str;
+}
+
+static inline void
+prof_log_prov(const struct fi_provider *prov)
+{
+	FI_TRACE(prov, FI_LOG_CORE, "\tprov: %s\n", prov->name);
+}
+
+static inline void
+prof_log_title(const struct fi_provider *prov, char *api_str, char *name)
+{
+	FI_TRACE(prov, FI_LOG_CORE, PROF_OUTPUT_FORMAT, api_str, name,
+             "Count", "Amount", "% Count", "% Amount");
+}
```

Comment:
The above 3 functions (ratio_to_str, prof_log_prov, prof_log_title) are 1-line functions that can just be expanded inline where called.  2 of these are only called from 1 place.